distribution/ltp-upstream/bin/ltp_log_parser.py

```diff
@@ -0,0 +1,307 @@
+
+
+def get_tc_name(l_tc):
+    tc_name = 'XXX-UNKNOWN-TEST-CASE'
+    for text in l_tc:
+        text = text.strip().rstrip()
+        if text.startswith(TC_NAME_FLAG):
+            tc_name = text.split(' ')[0].split('=')[1]
+            break
+    return tc_name
+
+
+def get_tc_result(l_tc):
+    tc_result = 'PASS'
+    for text in l_tc:
+        text = text.strip().rstrip()
+        if text.find(TC_RESULT_FAIL_FLAG) != -1:
+            tc_result = 'FAIL'
+            break
+    return tc_result
+
+
+def get_tc_metadata(tcno, l_tc):
+    d_tc = {}
+    d_tc['tcno'] = tcno
+    d_tc['name'] = get_tc_name(l_tc)
+    d_tc['result'] = get_tc_result(l_tc)
+    return d_tc
+
+
+def fmt_text(text):
+    width = TC_LOG_SUMMARY_WIDTH
+    nspaces = width - 4 - len(text)
+    return '# %s%s #' % (text, ' ' * nspaces)
+
+
+def print_summary(tcno, name, result):
+    width = TC_LOG_SUMMARY_WIDTH
+    print('#' * width)
+    print(fmt_text('Test Num    : %d' % tcno))
+    print(fmt_text('Test Case   : %s' % name))
+    print(fmt_text('Test Result : %s' % result))
+    print('#' * width)
+    print()
+
+
+def print_refined_text(l_tc, spaces):
+    l_dll = init_doubly_linked_list(l_tc)
+    omit_doubly_linked_list(l_dll)
+    l_out = traverse_doubly_linked_list(l_dll)
+    for line in l_out:
+        print('%s%s' % (spaces, line))
+
+
+def print_raw_text(l_tc, spaces):
+    lineno = 0
+    for line in l_tc:
+        lineno += 1
+        print('%s%6d\t%s' % (spaces, lineno, line))
+
+
+def dump(d_tc_metadata, l_tc):
+    tcno = d_tc_metadata['tcno']
+    name = d_tc_metadata['name']
+    result = d_tc_metadata['result']
+
+    print('>>> %d:%s %s <<<' % (tcno, name, result))
+    print_summary(tcno, name, result)
+    if TC_LOG_OUTPUT_RAW:
+        print_raw_text(l_tc, ' ' * TC_LOG_OUTPUT_INDENT)
+    else:
+        print_refined_text(l_tc, ' ' * TC_LOG_OUTPUT_INDENT)
+    print()
+
+
+def parse(file_handle):
+    l_tcs = []
+    while True:
+        line = file_handle.readline()
+        if not line:
+            break
+
+        line = line.strip().rstrip()
+        if line.strip().startswith(TC_START):
+            loop_flag = True
+            l_tc = []
+            l_tc.append(line)
+            continue
+
+        if line.strip().startswith(TC_END):
+            loop_flag = False
+            l_tc.append(line)
+            l_tcs.append(l_tc)
+            continue
+
+        if loop_flag:
+            l_tc.append(line)
+    return l_tcs
+
+
+def usage(prog):
+    sys.stderr.write('Usage: %s [-r] [-t indent] [-F] <logfile>\n' % prog)
+    sys.stderr.write('e.g.\n')
+    sys.stderr.write('o dump failed test cases\n')
+    sys.stderr.write('       %s -F /tmp/foo.log\n' % prog)
+    sys.stderr.write('o dump all test cases and remove those repeated lines\n')
+    sys.stderr.write('       %s /tmp/foo.log\n' % prog)
+    sys.stderr.write('o dump all test cases without removing repeated lines\n')
+    sys.stderr.write('       %s -r /tmp/foo.log\n' % prog)
```

Comment:
All of these `sys.stderr.write` should probably be replaced by `print(..., file=sys.stderr)`.

Reply:
Hi @emanchado, it's great to help to write the code of ltp_log_parser.py to be more Pythonic. To tell the truth, I don't have much experience writing Python code. I'm a C+bash programmer. When you are done, please send out, I'll help to review. Thanks again!

Reply:
Could you send me a log file so I can test, and make sure the output doesn't change between your version and my version?

Reply:
I have pushed a new commit to the branch. I have tested it yet, so it might be broken, but at least you can start having a look at what I had in mind.

When you send me a test file I'll check and will fix whatever is needed so the output doesn't change between my version and yours.

Reply:
Hi @emanchado, thank you for updating the Python script. It's great to use functions in an existing module to replace the doubly linked list I used. I am not familiar with `itertools`, please help to update the script again as it is broken.
* http://pastebin.test.redhat.com/822320
Test log file to test the Python script:
* http://10.66.13.186/t/sandbox/ark/PR449/foo.log